src/main/scala/is/hail/io/bgen/BgenRecord.scala

```diff
@@ -78,6 +78,73 @@ class BgenRecordV11(compressed: Boolean,
   }
 }
 
+class BGen12ProbabilityArray(a: Array[Byte], nSamples: Int, nGenotypes: Int, nBitsPerProb: Int) {
+
+  def apply(s: Int, i: Int): UInt = {
+    val firstBit = (s * (nGenotypes - 1) + i) * nBitsPerProb
+    val firstBitWithinByte = firstBit & 7
+
+    var j = nSamples + 10 + (firstBit >> 3)
+    var r = (a(j) & 0xff) >>> (firstBit & 7)
+    var k = 8 - (firstBit & 7)
+    j += 1
```

Comment:
Rename `j` and `k` to something more descriptive. Maybe `j` can be `byteIndex` and `k` can be `bufferSize`? `r` is okay -- I'm assuming that's for `result`.